### DISCUSSION TITLE
GeoJSON object no longer receives a default type

### DISCUSSION THREAD

**Discussion by Jeffrey-Vervoort-KNMI** [developmentseed/geojson-pydantic#128] [General]
After upgrading from `0.4.3` to `0.6.2` the validation raises an exception for `type` in `FeatureCollection`, `Feature` and `Point`, in our code but also for the [Usage example in the readme.md](https://github.com/developmentseed/geojson-pydantic/tree/main#usage).

At version `0.4.3` GeoJSON objects received a default see below:

```Python
class FeatureCollection(GenericModel, Generic[Geom, Props], GeoInterfaceMixin):
    """FeatureCollection Model"""
    type: str = Field("FeatureCollection", const=True)
```

The `type` field changed in `0.6.0` to a `Literal` and the default is removed:
```Python
class FeatureCollection(GenericModel, Generic[Geom, Props], GeoInterfaceMixin):
    """FeatureCollection Model"""
    type: Literal["FeatureCollection"]
```

When I read the [spec](https://datatracker.ietf.org/doc/html/rfc7946#section-3): `A GeoJSON object has a member with the name "type". The value of the member MUST be one of the GeoJSON types.`, I think that it should receive a default. If that would be the case then it could be changed to: 

```Python
class FeatureCollection(GenericModel, Generic[Geom, Props], GeoInterfaceMixin):
    """FeatureCollection Model"""
    type: Literal["FeatureCollection"] = "FeatureCollection"
```

If this would be an issue, then it would be necessary to also change it for the other types: `Feature` and `Point` as well.

**Comment by vincentsarago** [comment]
Hi @Jeffrey-Vervoort-KNMI 
Sorry for the frustration, we changed this knowing it will be quite a breaking change but we also wanted to make sure the library was following the specification to the letter. (https://github.com/developmentseed/geojson-pydantic/blob/main/CHANGELOG.md#060a0---2023-04-04) 

ref: https://github.com/developmentseed/geojson-pydantic/issues/92#issuecomment-1406426704

a Feature, FeatureCollection and Geometry dictionary should have a type. If not we cannot discriminate between them. We do think that this make the library more robust but we also understand that this was a big breaking change 

also already discussed in https://github.com/developmentseed/geojson-pydantic/discussions/124

Thanks for pointing out the error in the Readme, It will be fixed

**Comment by alexismanin** [comment]
@vincentsarago I am not really convinced about that justification.

I understand that an input json that we want to decode must provide the type attribute. But on the Python class, that's a whole different story.

 "type" is required in JSON, because the only way to define a "named type" (i.e. a class) in this format is by putting a property holding type name (in geojson, its the role of "type" attribute)  in the object itself.

In Python, this information is held by the class name itself.

Alas, when instantiating a Python object, you give the type through its class name. Therefore, there is not point to repeat it.
Moreover, with a default value, type attribute is still present and mandatory, as Pydantic instantiate the field with the default value, so it is not missing.

Also, validation is expressed through `Literal` typing, that checks input value is the expected literal. So, even with a default value, type remains mandatory and constant: if you do `Feature.model_validate({"type": "FeatureCollection"})`, you will get a validation error for type object that has the wrong value, even if you define a default value. And when you do : `Feature.model_validate({})`, you also get a validation error because the overral structure does not match the Feature model. 

In fact, the `type` attribute in geojson is here for polymorphism purpose, which is not applicable anyway with the current pydantic mapping, as each model defines independently a constant "type" attribute.

So, I think type should have a default/freezed value. It would:

 - avoid redundancy, which, in the current case, is really very annoying
 - Remain valid with the spec: any GeoJSON object will always have a type value set with the right value.
 - It does not impact JSON (de)serialization in any way.
